ENH: Adding AeroSurfaces classes

Gui-FernandesBR opened the pull request:
## Pull request type
- [X] Code maintenance (refactoring, formatting, renaming, tests)

## Pull request checklist

- Code base maintenance (refactoring, formatting, renaming):

  - [X] Docs have been reviewed and added / updated if needed
  - [X] Lint (`black rocketpy`) has passed locally and any fixes were made
  - [X] All tests (`pytest --runslow`) have passed locally

## What is the current behavior?

It is impossible to access arguments of an exisiting aerodynamic surface.

## What is the new behavior?

Following the Parachute.py file example, I'm changing the addTail, addFins, addNose to be working with the help of separate classes. This way, we can better capture data from each object. 

Also, no more dictionaries to save aerodynamic surfaces data, instead let's keep everything inside classes! `getattr(obj, attribute)` will make our life easier.

## Does this introduce a breaking change?
- [ ] Yes
- [X] No

## Other information
I'm doing this just to make our life easier on the future dispersion class

giovaniceotto commented:
What a great surprise @Gui-FernandesBR! Since this exceeds the limit of 200 line changes, I request a discord call so that we can discuss the work that has been done in this PR before review.

How about 2022/10/24, from 22:00 UTC until 22:30 UTC?

@MateusStano is, of course, also invited as he is a reviewer and the back-end manager.

Feel free to suggest another option if this does not work for you.

Gui-FernandesBR commented:
> What a great surprise @Gui-FernandesBR! Since this exceeds the limit of 200 line changes, I request a discord call so that we can discuss the work that has been done in this PR before review.
> 
> How about 2022/10/24, from 22:00 UTC until 22:30 UTC?
> 
> @MateusStano is, of course, also invited as he is a reviewer and the back-end manager.
> 
> Feel free to suggest another option if this does not work for you.

This time doens't work for me, let's try another way, I will contact you.
Please don't worry about the size of this PR, you should be worried about what comes next (the dispersion class!). 💔
For now, surviving tests and running notebooks without breaking the code are the easiest confirmations I think you could try

MateusStano commented:
Oh, one more thing. The `getting_started.ipynb` notebook is set with elliptical fins instead of trapezoidal like it used to be. Is this intended?

Gui-FernandesBR commented:
I will try to review this one by today, but I am already a fan of your commits.

I don't see any reason to use elliptical fins on the getting started notebook.

Indeed, omce the Calisto rocket presented trapezoidal fins, let's keep with those...

giovaniceotto commented on rocketpy/Flight.py:
```diff
@@ -1412,7 +1412,8 @@ def uDot(self, t, u, postProcessing=False):
                 if -1 * compStreamVzBn < 1:
                     compAttackAngle = np.arccos(-compStreamVzBn)
-                    cLift = aerodynamicSurface["cl"](compAttackAngle, freestreamMach)
+                    cLift = aerodynamicSurface.cl(compAttackAngle, freestreamMach)
+                    cLift = aerodynamicSurface.cl(compAttackAngle, freestreamMach)
```
```suggestion
                    cLift = aerodynamicSurface.cl(compAttackAngle, freestreamMach)
```

giovaniceotto commented on rocketpy/Flight.py:
```diff
@@ -1428,8 +1429,8 @@ def uDot(self, t, u, postProcessing=False):
             # Calculates Roll Moment
-            if aerodynamicSurface["name"] == "Fins":
-                Clfdelta, Cldomega, cantAngleRad = aerodynamicSurface["roll parameters"]
+            if aerodynamicSurface.name == "Fins":
+                Clfdelta, Cldomega, cantAngleRad = aerodynamicSurface.rollParameters
```
Should we apply EAFP here rather than LBYL?

https://realpython.com/python-lbyl-vs-eafp/

Gui-FernandesBR replied:
Yes, I think we should, let me make a suggestion here, and you can say if it is clear enough or not: 

```suggestion
            try:
                Clfdelta, Cldomega, cantAngleRad = aerodynamicSurface.rollParameters
            except AttributeError:
                pass
```

However, I think we now should do something different, cheking the type of the aerodynamic surface, what do you think? And for that *we should import Fins class in the beggining of the file*

```suggestion
            if isinstance(aerodynamicSurface, Fins):
                Clfdelta, Cldomega, cantAngleRad = aerodynamicSurface.rollParameters
````

giovaniceotto commented:
There are still some improvements to be made before this PR gets merged... but it is of mangnificent importance to RocketPy in the long run! Can't wait for this!

Gui-FernandesBR commented:
Hello @MateusStano , this is getting intersting... Really thank you for pushing such good features to these classes.

Describing what I did after your 1st review:
- [x] Added allInfo methods for both Tail and Nose 
- [x] Updated all the docstrings, being ready to go for a merge
- [x] Modify a single at Function to properly handle the 2D plots

My review comments are the following (6):

1. Can we analyze the name of some methods? Here is a list of what I saw we could change:
- `Fins.setAttribute` can create a confusion with `__setattr__` method, I suggest `modifyAttribute` or similar
- `evaluateRollCoefficients` could be `evaluateRollParameters`, since it mainly creates the rollParameters attribute
- `liftInfo` could be `aerodynamicInfo`?
- Also, any reason for keeping a single underscore at these methods? `_beta` and `_finNumCorrection`
- And just out of curiosity... In the future, will we use this beta function to other aero surfaces as well? 

3. I'd like to understand if these are correct: 
`self.rollParameters[0]()  # TODO: improve this plot`
`self.rollParameters[1]()  # TODO: improve this plot`
![image](https://user-images.githubusercontent.com/63590233/199427477-2c2cd885-fd27-4d33-a596-13239ad5fc9c.png)
![image](https://user-images.githubusercontent.com/63590233/199427516-9f25c00f-8996-48ff-b043-75f7d49ec413.png)

4. The Fin.draw() method is good, but I think we can move forward without this functionality for Tail and Nose. In the future we might want to do a overkill draw method that takes the axes from each of theses drawings and create the rocket draw at once.
5. Regarding the NoseCone, why don't we take into considerations the nose radius/diameter? It took me years to realize that this is not an input for nose definition.
6. Maybe we should add these new allInfo methods to our getting started, or do you think we can accelerate and deliver a rocket_class_usage with this detailed information? Probably would be too many information at the getting started!

Gui-FernandesBR commented:
Also, I always like to give some demos here, so people can easily preview what is going on, so let's go:
- First, NoseCone.allInfo() method:
![image](https://user-images.githubusercontent.com/63590233/199427190-5a21f1c9-13e4-401a-8057-e813daee2620.png)

- Now, take a look on what Fins.allInfo can do:
![image](https://user-images.githubusercontent.com/63590233/199427255-b97e17d7-cf23-4a89-80c4-561590bd66a8.png)
![image](https://user-images.githubusercontent.com/63590233/199427312-d20707a0-2fbf-485e-94a8-e3ef9ee4776e.png)
![image](https://user-images.githubusercontent.com/63590233/199427355-0462d9dd-b407-4959-b385-b8d8a3ac4b31.png)
![image](https://user-images.githubusercontent.com/63590233/199427396-9c50705d-a3eb-40a9-a1a1-950c852955ae.png)
![image](https://user-images.githubusercontent.com/63590233/199427595-bab4aade-b555-4997-b39e-a9927f17d32d.png)
![image](https://user-images.githubusercontent.com/63590233/199427678-2de3b8b6-2375-47b6-a46c-4ba3491e4944.png)


- Finally, Tail.allInfo() is here:
![image](https://user-images.githubusercontent.com/63590233/199427801-75ad70a5-97c1-4f61-b1f9-3d1b1cbde769.png)

Gui-FernandesBR commented:
Question: 
- Should we implement drag coefficient based on both Mach and alpha? As far as I know, both openrocket and Rasaero can create that. This made me to rethink my answer on this issue: https://github.com/RocketPy-Team/RocketPy/issues/46

Gui-FernandesBR commented:
TODO: 
- transform def Fins(self) in an abstract class 
- AerodynamicSurface() also having an abstract class

- Add setAttribute ...
-
